lib/helpers.py

```diff
@@ -10,437 +10,598 @@
+            return self._from_parquet_corrected(
+                file_name=cache_file_name,
+                s3_file_name=s3_cache_file_name,
+                fs=fs,
+                columns=columns,
+            )
 
-    # write it to the S3 cache folder as well
-    print(datetime.now(), 'caching on S3 as parquet', s3_cache_filename)
-    to_parquet(df, s3_cache_filename)
+        if fs.exists(path=s3_cache_file_name):
+            print(now, 'loading from S3 cache', s3_cache_file_name)
 
-    return df
+            # Download the file to local cache first to avoid timeouts during the load.
+            # This way, if they happen, restart will be using local copies first.
+            fs.get(s3_cache_file_name, cache_file_name)
 
+            print(now, 'stored S3 cache file to local drive, loading', cache_file_name)
 
-def calculate_ad_spend(df):
-    ad_spend_micros = df[(df.event_type == 'buying_conversion') & (df.ab_test_group == TEST)]['cost_eur'].sum()
-    return ad_spend_micros / 10 ** 6
+            return self._from_parquet_corrected(
+                file_name=cache_file_name,
+                s3_file_name=s3_cache_file_name,
+                fs=fs,
+                columns=columns,
+            )
 
+        print(now, 'start loading CSV for', audience, source, date)
 
-def marked(df):
-    """
-    The dataframe created by `marked` will contain all mark events. Remerge marks users per campaign. If a user was
-    marked once for an audience he will have the same group allocation for consecutive marks (different campaigns)
-    unless manually reset on audience level.
-    """
-    mark_df = df[df.event_type == 'mark']
+        read_csv_kwargs = {'chunksize': self.chunk_size}
+        if columns:
+            read_csv_kwargs['usecols'] = columns
 
-    # we dont need the event_type anymore (to save memory)
-    mark_df = mark_df.drop(columns=['event_type'])
+        df = pd.DataFrame()
 
-    sorted_mark_df = mark_df.sort_values('ts')
+        if not fs.exists(path=filename):
+            print(now, 'WARNING: no CSV file at for: ', audience, source, date, ', skipping the file: ', filename)
+            return df
 
-    deduplicated_mark_df = sorted_mark_df.drop_duplicates(['user_id'])
+        for chunk in pd.read_csv(filename, escapechar='\\', low_memory=False, **read_csv_kwargs):
+            if chunk_filter_fn:
+                filtered_chunk = chunk_filter_fn(chunk, self.revenue_event)
+            else:
+                filtered_chunk = chunk
 
-    return deduplicated_mark_df
+            if source != CSV_SOURCE_ATTRIBUTIONS:
+                # we are not interested in events that do not have a group amongst non-attribution events
+                filtered_chunk = filtered_chunk[filtered_chunk['ab_test_group'].isin(['test', 'control'])]
 
+            # remove events without a user id
+            filtered_chunk = filtered_chunk[filtered_chunk['user_id'].str.len() == 36]
```

Comment:
The value 36 is used 3 times in the code, maybe it's worth to set a constant?

Reply:
:+1: